File: backend/src/CarWash.Domain/Entities/Filial.cs
using System.Text.RegularExpressions;
using CarWash.Domain.Common;
using CarWash.Domain.ValueObjects;

namespace CarWash.Domain.Entities;

/// <summary>
/// Unidade operacional (RF017/RF018). Capacidade controlada por
/// <c>celulas_ativas BETWEEN 1 AND 100</c> (RN009 / CHECK <c>ck_filiais_celulas_faixa</c>).
/// Identificação operacional via <c>Codigo</c> (regex <c>^[A-Z0-9]{2,20}$</c>),
/// CNPJ e endereço estruturado opcionais durante o rollout aditivo do RF017.
/// </summary>
public sealed class Filial : IAuditable, IAuditableSetter
{
    public const int MinCelulasAtivas = 1;
    public const int MaxCelulasAtivas = 100;
    public const int NomeMinChars = 3;
    public const int NomeMaxChars = 120;
    public const int CodigoMinChars = 2;
    public const int CodigoMaxChars = 20;

    private static readonly Regex CodigoRegex = new(
        "^[A-Z0-9]{2,20}$",
        RegexOptions.Compiled | RegexOptions.CultureInvariant,
        TimeSpan.FromMilliseconds(50));

    private Filial()
    {
        Nome = null!;
        Codigo = null!;
        Timezone = null!;
    }

    public Guid Id { get; private set; }

    public string Nome { get; private set; }

    /// <summary>
    /// Identificador operacional curto da filial (RF017). Único case-sensitive
    /// no banco (UK parcial <c>uk_filiais_codigo</c> + CHECK
    /// <c>ck_filiais_codigo_formato</c>). Obrigatório no domínio mesmo que a
    /// coluna esteja nullable durante o rollout aditivo (ADR 0007 §3.4).
    /// </summary>
    public string Codigo { get; private set; }

    /// <summary>
    /// CNPJ opcional (L2 do ADR 0007). Quando presente, único (UK parcial
    /// <c>uk_filiais_cnpj</c>). Armazenado em 14 dígitos sem máscara.
    /// </summary>
    public string? Cnpj { get; private set; }

    public bool Ativa { get; private set; }

    public int CelulasAtivas { get; private set; }

    public string Timezone { get; private set; }

    public string? EnderecoCep { get; private set; }

    public string? EnderecoLogradouro { get; private set; }

    public string? EnderecoNumero { get; private set; }

    public string? EnderecoComplemento { get; private set; }

    public string? EnderecoBairro { get; private set; }

    public string? EnderecoCidade { get; private set; }

    public string? EnderecoUf { get; private set; }

    public DateTime CriadoEm { get; private set; }

    public DateTime AtualizadoEm { get; private set; }

    /// <summary>
    /// Id do usuário autenticado que criou a filial (auditoria — espelha
    /// <c>Cliente.CriadoPorUsuarioId</c>). Nullable para tolerar registros
    /// legados anteriores à migration <c>AdicionaCadastroFilial</c>.
    /// </summary>
    public Guid? CriadoPorUsuarioId { get; private set; }

    /// <summary>
    /// Getter computado equivalente ao VO <see cref="Endereco"/>. Retorna
    /// <c>null</c> quando o endereço estruturado não foi informado (rollout
    /// aditivo). Não persistido — <c>FilialConfiguration</c> aplica
    /// <c>Ignore(x =&gt; x.Endereco)</c>.
    /// </summary>
    public Endereco? Endereco
    {
        get
        {
            // Durante o rollout aditivo, as colunas de endereço podem existir de forma parcial.
            // Só materializa o VO quando todos os campos obrigatórios estiverem presentes.
            if (string.IsNullOrWhiteSpace(EnderecoCep)
                || string.IsNullOrWhiteSpace(EnderecoLogradouro)
                || string.IsNullOrWhiteSpace(EnderecoNumero)
                || string.IsNullOrWhiteSpace(EnderecoBairro)
                || string.IsNullOrWhiteSpace(EnderecoCidade)
                || string.IsNullOrWhiteSpace(EnderecoUf))
            {
                return null;
            }

            return new Endereco(
                EnderecoCep,
                EnderecoLogradouro,
                EnderecoNumero,
                EnderecoComplemento,
                EnderecoBairro,
                EnderecoCidade,
                EnderecoUf);
        }
    }

    /// <summary>
    /// Fábrica do agregado <see cref="Filial"/> (RF017 + RF018). Aplica as
    /// invariantes do domínio antes da persistência. <paramref name="codigo"/>
    /// é obrigatório (regra de negócio crítica, ADR 0007 §2.4) — mesmo que a
    /// coluna esteja nullable em produção durante o rollout aditivo.
    /// </summary>
    public static Filial Criar(
        Guid id,
        string nome,
        string codigo,
        int celulasAtivas,
        Endereco? endereco = null,
        Cnpj? cnpj = null,
        string? timezone = null)
    {
        if (id == Guid.Empty)
        {
            throw new DomainException("Id da filial não pode ser vazio.");
        }

        if (string.IsNullOrWhiteSpace(nome)
            || nome.Trim().Length < NomeMinChars
            || nome.Trim().Length > NomeMaxChars)
        {
            throw new DomainException(
                $"Nome da filial deve ter entre {NomeMinChars} e {NomeMaxChars} caracteres.");
        }

        if (string.IsNullOrWhiteSpace(codigo) || !CodigoRegex.IsMatch(codigo))
        {
            throw new DomainException(
                "Código da filial é obrigatório e deve conter de 2 a 20 caracteres alfanuméricos maiúsculos (A-Z, 0-9).");
        }

        if (celulasAtivas is < MinCelulasAtivas or > MaxCelulasAtivas)
        {
            throw new DomainException(
                $"Células ativas deve estar entre {MinCelulasAtivas} e {MaxCelulasAtivas} (RN009).");
        }

        var agora = DateTime.UtcNow;
        return new Filial
        {
            Id = id,
            Nome = nome.Trim(),
            Codigo = codigo,
            Cnpj = cnpj?.Valor,
            CelulasAtivas = celulasAtivas,
            Timezone = string.IsNullOrWhiteSpace(timezone) ? "America/Sao_Paulo" : timezone,
            Ativa = true,
            EnderecoCep = endereco?.Cep,
            EnderecoLogradouro = endereco?.Logradouro,
            EnderecoNumero = endereco?.Numero,
            EnderecoComplemento = endereco?.Complemento,
            EnderecoBairro = endereco?.Bairro,
            EnderecoCidade = endereco?.Cidade,
            EnderecoUf = endereco?.Uf,
            CriadoEm = agora,
            AtualizadoEm = agora,
        };
    }

    public void AjustarCelulas(int novoValor)
    {
        if (novoValor is < MinCelulasAtivas or > MaxCelulasAtivas)
        {
            throw new DomainException(
                $"Células ativas deve estar entre {MinCelulasAtivas} e {MaxCelulasAtivas} (RN009).");
        }

        CelulasAtivas = novoValor;
    }

    public void Inativar() => Ativa = false;

    public void Ativar() => Ativa = true;

    /// <summary>
    /// Registra o usuário responsável pela criação (auditoria — espelha
    /// <c>Cliente.RegistrarCriadoPor</c>). Chamado pelo handler ao persistir
    /// uma filial nova.
    /// </summary>
    public void RegistrarCriadoPor(Guid? usuarioId)
    {
        CriadoPorUsuarioId = usuarioId;
    }

    void IAuditableSetter.SetCriadoEm(DateTime valor) => CriadoEm = valor;

    void IAuditableSetter.SetAtualizadoEm(DateTime valor) => AtualizadoEm = valor;
}

File: frontend/src/types/filial.ts
/**
 * Tipos de filial (RF017/RF018/RF019).
 *
 * <p>Cobrem dois usos: a listagem/gerência de filiais (cadastro e consulta —
 * `GET/POST /api/v1/filiais`) e o resumo consumido pelo seletor obrigatório do
 * agendamento (RF019/RN010, `GET /api/v1/filiais?ativo=true`).</p>
 *
 * <p>O contrato segue o backend (ADR-0007 §4): o resumo da listagem expõe
 * apenas `id, nome, codigo, cidade, uf, ativo`; o endereço no cadastro é um
 * objeto estruturado (`EnderecoFilialRequest`).</p>
 */

export interface FilialResumo {
  id: string;
  nome: string;
  /** Código/identificador operacional da filial (exibido junto ao nome — RF019). */
  codigo?: string;
  cidade?: string;
  uf?: string;
  ativo: boolean;
}

export interface ListaFiliais {
  itens: FilialResumo[];
  total: number;
}

/**
 * Endereço estruturado da filial — espelha `EnderecoFilialRequest` no backend.
 *
 * <p>Quando enviado, todos os campos (exceto `complemento`) são obrigatórios.
 * `cep` vai apenas com dígitos.</p>
 */
export interface EnderecoFilialRequest {
  cep: string;
  logradouro: string;
  numero: string;
  complemento?: string | null;
  bairro: string;
  cidade: string;
  uf: string;
}

/**
 * Payload de cadastro de filial — `POST /api/v1/filiais`.
 *
 * <p>`cnpj` é enviado apenas com dígitos (sem máscara) e omitido (`null`)
 * quando não informado. `endereco` é um objeto estruturado.</p>
 */
export interface CriarFilialRequest {
  nome: string;
  codigo: string;
  cnpj?: string | null;
  celulasAtivas: number;
  timezone?: string | null;
  endereco: EnderecoFilialRequest;
}

/** Envelope canônico do `POST /api/v1/filiais` (ADR-0007 §4.1). */
export interface CriarFilialResponse {
  id: string;
  mensagem: string;
  traceId: string;
}

/**
 * Detalhe da filial — `GET /api/v1/filiais/{id}` e resposta do
 * `PATCH /api/v1/filiais/{id}/celulas-ativas` (`FilialResponse` no backend).
 *
 * <p>Atenção ao contrato: este detalhe expõe apenas nome, capacidade, timezone,
 * status e auditoria — não inclui código, CNPJ nem endereço (RF018).</p>
 */
export interface FilialDetalhe {
  id: string;
  nome: string;
  celulasAtivas: number;
  timezone: string;
  ativa: boolean;
  criadoEm: string;
  atualizadoEm: string;
}
